fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/WorldRenderContext.java

```diff
@@ -0,0 +1,203 @@
+		 * must be drawn directly to the frame buffer, preferably in {@link WorldRenderEvents#LAST} to avoid being
+		 * overdrawn or cleared.
+		 */
+		VertexConsumerProvider consumers();
+	}
+
+	/**
+	 * Conveys WorldRenderer state to event handlers after invocation.
+	 * Includes view frustum, which is not available then.
+	 */
+	public interface AfterFrustum {
+		Frustum frustum();
+	}
+
+	/**
+	 * Used in the block breaking callback to convey information on
+	 * ray trace hit and accept cancellation of default outline render.
+	 */
+	public interface PreBlockOutline {
+		/**
+		 * The game object currently under the crosshair target. Normally
+		 * equivalent to {@link MinecraftClient#crosshairTarget}.
+		 * Provided for convenience.
+		 */
+		@Nullable HitResult hitResult();
+
+		/**
+		 * Call to prevent the default block outline from rendering.
+		 * Use this for event handlers that fully replace the vanilla
+		 * block outline render.
+		 *
+		 * <p>The effect of this method depends on the event from which it is called:
+		 *
+		 * <p>In {@code BEFORE_BLOCK_RENDER} this prevents invocation of
+		 * the {@code BLOCK_RENDER} event and {@link AfterBlockOutline#didCancelDefaultBlockOutline()}
+		 * will be {@code false} in later-stage event invocations. It (obviously)
+		 * also prevents the default outline render from running.
+		 *
+		 * <p>In {@code BLOCK_RENDER} this prevents the default outline render
+		 * from running and {@link AfterBlockOutline#didCancelDefaultBlockOutline()}
+		 * will be {@code true} in later-stage event invocations.
+		 *
+		 * <p>In both cases, this method is idempotent - calling it from more
+		 * than one event subscriber has the same effect as calling it once.
+		 * All event handlers within the same event invocation will always
+		 * execute and the cancellation status cannot be inspected during the
+		 * block outline render events.
+		 */
+		void cancelDefaultBlockOutline();
+	}
+
+	/**
+	 * Used in {@code BLOCK_OUTLINE} to convey the parameters normally sent to
+	 * {@code WorldRenderer.drawBlockOutline}.
+	 */
+	public interface BlockOutline {
+		VertexConsumer vertexConsumer();
+
+		Entity entity();
+
+		double cameraX();
+
+		double cameraY();
+
+		double cameraZ();
+
+		BlockPos blockPos();
+
+		BlockState blockState();
+	}
+
+	/**
+	 * Used after the block outline events to test if {@link #cancelDefaultBlockOutline()}
+	 * was called during the {@code BLOCK_OUTLINE} event.
+	 */
+	public interface AfterBlockOutline {
+		/**
+		 * Tests if the default block outline render was cancelled in the {@code BLOCK_OUTLINE} event.
+		 *
+		 * <p>Will NOT be true if the outline was cancelled during {@code BEFORE_BLOCK_OUTLINE}.
+		 *
+		 * @return {@code true} if the default block outline render was cancelled in the {@code BLOCK_OUTLINE} event.
+		 */
+		boolean didCancelDefaultBlockOutline();
+	}
+
+	public interface FrustumContext extends WorldRenderContext, AfterFrustum { }
```

Comment:
I feel the `Context` on the end of these nested interfaces may be redundant. At the same time `WorldRenderContext.Frustrum` could be a bit weird, so this needs a bit of an evaluation.

Reply:
Not a hill I'm prepared to die on, but absent some specific direction leaving as is.